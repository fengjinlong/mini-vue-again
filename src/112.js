function pathKeyedChildren(c1, c2, container, parentComponent, parentAnthor) {
  // ...

  if (i > e1) {
    // 第三种情况
  } else if (i > e2) {
    // 第四种情况
  } else {
    let s1 = i;
    let s2 = i;
    const toBePatched = e2 - s2 + 1;
    let patched = 0;

    // 映射关系
    const keyToNewIndexMap = new Map();

    // 新的映射关系
    for (let i = s2; i <= e2; i++) {
      const nextChild = c2[i];
      keyToNewIndexMap.set(nextChild.key, i);
    }

    // 老的映射关系
    for (let i = s1; i <= e1; i++) {
      // 老节点 prevChild
      const prevChild = c1[i];
      if (patched >= toBePatched) {
        // 新的已经对比完，但是老的还没完事。直接删除
        hostRemove(prevChild.el);
        // 进入下一次循环
        continue;
      }
      let newIndex;
      /**
       *  如果 newIndex 存在，说明 prevChild 在新的里面存在。
       *  如果用户写了key，用key映射查找。如果没写key,用循环查找
       */
      if (prevChild.key !== null) {
        newIndex = keyToNewIndexMap.get(prevChild.key);
      } else {
        for (let j = s2; j <= e2; j++) {
          if (isSameNodeType(c2[j], prevChild)) {
            newIndex = j;
            break;
          }
        }
      }

      if (newIndex === undefined) {
        // 说明不存在prevChild，删掉老的 prevChild
        hostRemove(prevChild.el);
      } else {
        newIndexToOldIndexMap[newIndex - s2] = i + 1;
        // 存在，继续进行深度对比
        patch(prevChild, c2[newIndex], container, parentComponent, null);
      }
    }
  }
}
