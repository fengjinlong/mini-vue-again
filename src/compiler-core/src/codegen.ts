export function generate(ast) {
  const context = createCodegenContext();
  const { push } = context;

  console.log(ast);
  // 导入逻辑 const { toDisplayString: _toDisplayString } = Vue
  genFunctionPreamble(ast, context);

  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(", ");
  push(`function ${functionName}(${signature}) {`);
  push("return ");
  genNode(ast.codegenNode, context);
  push("}");
  return {
    code: context.code,
  };
}
function genFunctionPreamble(ast, context) {
  const { push } = context;
  const VueBinging = "Vue";
  console.log("ast---");
  console.log(ast.helpers.length);
  if (ast.helpers.length) {
    const aliasHelper = (s) => `${s}: _${s}`;
    push(
      `const { ${ast.helpers.map(aliasHelper).join(", ")} } = ${VueBinging}`
    );
    push("\n");
  }
  push("return ");
}

function genNode(node: any, context) {
  const { push } = context;
  push(`'${node.content}'`);
}

function createCodegenContext() {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
  };
  return context;
}
