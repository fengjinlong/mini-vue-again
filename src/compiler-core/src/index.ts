// 每个模块基于 此 对外导出方法
export * from "./compile";
