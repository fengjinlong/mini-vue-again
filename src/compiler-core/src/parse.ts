export function baseParse(content: string) {
  const context = createParserContext(content);

  return createRoot(parseChildren(context));
}

function parseChildren(context) {
  const nodes: any = [];
  const node = parseInterpolation(context);
  nodes.push(node);
  return nodes;
}

function parseInterpolation(context: any) {
  // {{xxx}}
  const closeIndex = context.source.indexOf("}}", 2);
  context.source = context.source.slice(2);
  const rawContentLength = closeIndex - 2;
  const content = context.source.slice(0, rawContentLength);

  // delete }}
  context.source = context.source.slice(rawContentLength, 2);

  return {
    type: "interpolation",
    content: {
      type: "simple_expression",
      content: content,
    },
  };
}

function createRoot(children) {
  return {
    children,
  };
}
function createParserContext(context: any) {
  return {
    source: context,
  };
}
