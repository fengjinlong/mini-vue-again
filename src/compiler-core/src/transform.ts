import { NodeTypes } from "./ast";

export function transform(root, options = {}) {
  // 全局上下文
  const context = createTransformContext(root, options);
  // 遍历
  traverseNode(root, context);
  createRootCodegen(root);
  root.helpers = [...context.helpers.keys()];
  // 修改
}
function createRootCodegen(root) {
  root.codegenNode = root.children[0];
}
function traverseNode(node: any, context) {
  const nodeTransformer = context.nodeTransformer;
  for (let i = 0; i < nodeTransformer.length; i++) {
    let transform = nodeTransformer[i];
    transform(node);
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      // 如果是插值 需要 toDisplayString
      context.helper("toDisplayString");
      break;
    default:
      break;
  }
  traverseChildren(node, context);
}
function traverseChildren(node: any, context: any) {
  const children = node.children;
  if (children) {
    for (let i = 0; i < children.length; i++) {
      const node = children[i];
      traverseNode(node, context);
    }
  }
}

function createTransformContext(root: any, options: any) {
  const context = {
    root,
    nodeTransformer: options.nodeTransformer || {},
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1);
    },
  };
  return context;
}
