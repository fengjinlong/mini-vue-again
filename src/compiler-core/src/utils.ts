import { NodeTypes } from "./ast";

export function isText(node) {
  return (node.nodeType =
    NodeTypes.TEXT || node.nodeType === NodeTypes.INTERPOLATION);
}
