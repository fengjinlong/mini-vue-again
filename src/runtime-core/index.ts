export {createApp} from './createApp'
export {h} from './h'
export {renderSlots} from "./helpers/renderSolts"
export {createTextVNode} from "./vnode"