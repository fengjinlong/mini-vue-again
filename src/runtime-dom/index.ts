import { createRenderer } from "../runtime-core";
function createElement(type: string) {
  // console.log("dom ------api")
  return document.createElement(type);
}
function patchProps(el, key, val) {
  // console.log("dom ------api")
  const isOn = (key: string) => /^on[A-Z]/.test(key);
  if (isOn(key)) {
    const event = key.slice(2).toLowerCase();
    el.addEventListener(event, val);
  } else {
    el.setAttribute(key, val);
  }
}
function insert(el, parent) {
  // console.log("dom ------api")
  parent.append(el);
}

const renderer: any= createRenderer({
  createElement,
  patchProps,
  insert,
});

// return {
//   createApp: createAppAPI(render)
// }
export function createApp(...args) {
  return renderer.createApp(...args);

  // 调用流程
  // return createAppAPI(render)(...args);
  // export function createAppAPI(render) {
  //   return function createApp(rootComponent) {
  //     return {
  //       mount(rootContainer) {
  //         // 先创建 vnode
  //         // component -> vnode
  //         // 所有逻辑操作 都会基于 vnode 做处理
  //         const vnode = createVNode(rootComponent);
  //         // 渲染虚拟节点
  //         render(vnode, rootContainer);
  //       },
  //     };
  //   }
  // }
}

export * from "../runtime-core"
export * from "../reactivity"