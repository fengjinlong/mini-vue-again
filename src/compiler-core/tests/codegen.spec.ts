import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";

describe("codegen", () => {
  it("string", () => {
    const ast = baseParse("hi");
    transform(ast)
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("interpolation", () => {
    const ast = baseParse("{{message}}");
    transform(ast, {
      nodeTransforms: [],
      // nodeTransforms: [transformExpression],
    });
    const { code } = generate(ast);
    // 快照
    expect(code).toMatchSnapshot();
  });
});
