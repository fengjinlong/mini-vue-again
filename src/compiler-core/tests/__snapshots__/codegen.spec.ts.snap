// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache) {return '[object Object]'}"
`;

exports[`codegen string 1`] = `"return function render(_ctx, _cache) {return 'hi'}"`;
